### PR TITLE
[FEGA usecase] merge `finalize` from `sda-pipeline` into `sda`

### PR DESCRIPTION
**Pull request by jbygdell**
* Merge `finalize` into `sda`
* Expands test suite to assign stable ID to a verified file.
* Adds backup functionality to finalize
  * As per the FEGA specification, an archived fie should be `backed up` before it can be considered `completed`
 * Adds check in the integration test that the files have been copied to the backup location.

### CODE REVIEW

File: sda/internal/database/db_functions.go [hunk 1]
```diff
@@ -288,3 +290,65 @@ func (dbs *SDAdb) getArchived(user, filepath, checksum string) (string, int, err
...
+		}
+		time.Sleep(time.Duration(math.Pow(2, float64(count))) * time.Second)
+	}
+
+	return exists, err
+}
+func (dbs *SDAdb) checkAccessionIDExists(accessionID string) (bool, error) {
+	dbs.checkAndReconnectIfNeeded()
+	db := dbs.DB
+	const checkIDExist = "SELECT COUNT(*) FROM sda.files WHERE stable_id = $1;"
```
Comment (MalinAhlberg) [inline sda/internal/database/db_functions.go:312]: out of curiosity, should the `select` on line 282 still be from `local_ega`?

**Review by MalinAhlberg — COMMENTED** [review]
Looks like solid work! And really nice with updated documentation!

I added some questions/comments on some places.  My suggested additions to the documentation might be a bit out of place, but I wanted to noted the places where I got stuck when reading.

**Comment by jbygdell** [comment]
> Looks like solid work! And really nice with updated documentation!
> 
> I added some questions/comments on some places. My suggested additions to the documentation might be a bit out of place, but I wanted to noted the places where I got stuck when reading.

Generally speaking the documentation needs an overhaul in order to get it up to date, for all services. But that might be easiest to do once this first batch of mere PRs are done.

File: sda/internal/database/db_functions.go (hunk 1 above)
Comment (jbygdell) [reply sda/internal/database/db_functions.go:311]: Of course it should not be.

File: sda/cmd/finalize/finalize.md [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+  - `info`
+  - `warn` (or `warning`)
+  - `error`
+  - `fatal`
+  - `panic`
+
+#### Keyfile settings
+
+These settings control which crypt4gh keyfile is loaded.
+These settings are only needed is `copyheader` is `true`.
```
[the remark this answers is not in the document]
Comment (jbygdell) [reply sda/cmd/finalize/finalize.md:79]: Copy header isn't used by `finalize` so this should be removed.

**Review by blankdots — COMMENTED** [review]
> Adds backup functionality to finalize 

so no more backup service, :thinking: should we remove that  ?

File: sda/internal/config/config.go [hunk 3]
```diff
@@ -356,22 +335,12 @@ func NewConfig(app string) (*Config, error) {
...
-
-		err := c.configBroker()
-		if err != nil {
-			return nil, err
-		}
-
-		err = c.configDatabase()
-		if err != nil {
-			return nil, err
+	case "finalize":
```
Comment (blankdots) [inline sda/internal/config/config.go:338]: i don't see a backup service any more maybe we should remove it with this PR if it is removed in config

Comment (jbygdell) [reply sda/internal/config/config.go:338]: The functionality of the backup service is needed in BigPicture together with the corresponding API component in order to replicate datasets between the two sites, although under a different name so people can't mix it up.
Once everything related FEGA is done I'll create the BP speciffic PRs.

Comment (blankdots) [reply sda/internal/config/config.go:338]: I see, ok.

File: sda/cmd/finalize/finalize.md [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+4. If the type of the `DecryptedChecksums` field in the message is `sha256`, the value is stored.
+5. A new RabbitMQ "complete" message is created and validated against the "ingestion-completion" schema. If the validation fails, an error message is written to the logs.
+6. The file accession ID in the message is marked as "ready" in the database. On error the service sleeps for up to 5 minutes to allow for database recovery, after 5 minutes the message is Nacked, re-queued and an error message is written to the logs.
+7. The complete message is sent to RabbitMQ. On error, a message is written to the logs.
+8. The original RabbitMQ message is Ack'ed.
+
+## Communication
+
+- Finalize reads messages from one rabbitmq queue (default `accessionIDs`).
+- Finalize writes messages to one rabbitmq queue (default `backup`).
```
Comment (blankdots) [inline sda/cmd/finalize/finalize.md:124]: from what i understand this is not the case

File: sda/cmd/finalize/finalize.md [hunk 5]
```diff
@@ -0,0 +1,125 @@
+# sda-pipeline: finalize
+
+Handles the so-called _Accession ID (stable ID)_ to filename mappings from Central EGA.
```
Comment (blankdots) [inline sda/cmd/finalize/finalize.md:3]: well not  necessarily for Central EGA

**Review by blankdots — APPROVED** [review]
small notes about documentation, but I think that will be resolved later so I approve as is.

**Comment by aaperis** [comment]
This looks nice and I understand that it required some work. But it should have been two PRs, one merging `finalize` to `sda-pipeline` and then another one addressing `backup`. 
Preferably, the fate of `backup` and its functionalities should have been written down on a ticket on some board during planning first.

**Review by blankdots — APPROVED** [review]

**Pull request merged**